HARP-8111 Refactored tile loader to fix bug and remove unecessary tile reloads.

MPursche opened the pull request:
Signed-off-by: Marcel Pursche marcel.pursche@here.com

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

nzjony commented on @here/harp-mapview-decoder/lib/TileLoader.ts:
```diff
@@ -104,25 +104,19 @@ export class TileLoader {
             case TileLoaderState.Loading:
+            case TileLoaderState.Loaded:
+            case TileLoaderState.Decoding:
+                // tile is already loading
```
it could also be Decoding

nzjony commented on @here/harp-mapview-decoder/lib/TileLoader.ts:
```diff
@@ -104,25 +104,19 @@ export class TileLoader {
     loadAndDecode(): Promise<TileLoaderState> {
         switch (this.state) {
             case TileLoaderState.Loading:
+            case TileLoaderState.Loaded:
```
I looked into the code, it seems that the Loaded state is set for a very short time period. Maybe it would be worth at least adding a comment that this is potentially no longer necessary.

nzjony commented on @here/harp-mapview-decoder/lib/TileLoader.ts:
```diff
@@ -104,25 +104,19 @@ export class TileLoader {
-                return this.donePromise!;
-
-            case TileLoaderState.Decoding:
-                this.cancelDecoding();
```
I'm not sure this is exactly correct, if you are decoding a Tile, and then call loadAndDecode, it probably still makes sense to cancel the request. For example, if you are changing styles, and you don't cancel the current request, you will get tiles with different styles.

MPursche replied:
When the style is changed the tiles of a datasource are marked as dirty and evicted from the cache. This also cancels running requests. There still seems to be a bug where tiles are displayed with different styles but it can also be triggered with the changes in this pull request.

nzjony replied:
You're right, ok cool.

MPursche commented:
The main issue is that in `TileLoader.ensureLoadingStarted()` the `TileLoaderState.Failed` isn't handled. So if loading of a tile fails once(i.e. no network connection), it will never be loaded, because the `TileLoader` can't recover from the failed state.
I refactored the code and removed  `TileLoader.ensureLoadingStarted()` because it is only called from `TileLoader.loadAndDecode()` and it doesn't make sense to check the state of the `TileLoader` in both methods. It makes more sense to only check the state in `loadAndDecode()` and handle the different branches there.

nzjony commented:
@heremaps-bot retest this please
